core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java

```diff
@@ -129,46 +136,75 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
+        partitionWriters[i] =
+                blockManager.getDiskWriter(blockId, file, serInstance, fileBufferSize, writeMetrics);
+      }
 
-    while (records.hasNext()) {
-      final Product2<K, V> record = records.next();
-      final K key = record._1();
-      partitionWriters[partitioner.getPartition(key)].write(key, record._2());
-    }
+      // Creating the file to write to and creating a disk writer both involve interacting with
+      // the disk, and can take a long time in aggregate when we open many files, so should be
+      // included in the shuffle write time.
+      writeMetrics.incWriteTime(System.nanoTime() - openStartTime);
 
-    for (int i = 0; i < numPartitions; i++) {
-      try (DiskBlockObjectWriter writer = partitionWriters[i]) {
-        partitionWriterSegments[i] = writer.commitAndGet();
+      while (records.hasNext()) {
+        final Product2<K, V> record = records.next();
+        final K key = record._1();
+        partitionWriters[partitioner.getPartition(key)].write(key, record._2());
       }
-    }
 
-    File output = shuffleBlockResolver.getDataFile(shuffleId, mapId);
-    File tmp = Utils.tempFileWith(output);
-    try {
-      partitionLengths = writePartitionedFile(tmp);
-      shuffleBlockResolver.writeIndexFileAndCommit(shuffleId, mapId, partitionLengths, tmp);
-    } finally {
-      if (tmp.exists() && !tmp.delete()) {
-        logger.error("Error while deleting temp file {}", tmp.getAbsolutePath());
+      for (int i = 0; i < numPartitions; i++) {
+        try (DiskBlockObjectWriter writer = partitionWriters[i]) {
+          partitionWriterSegments[i] = writer.commitAndGet();
+        }
+      }
+
+      File output = shuffleBlockResolver.getDataFile(shuffleId, mapId);
+      File tmp = Utils.tempFileWith(output);
+      try {
+        partitionLengths = writePartitionedFile(tmp);
+        shuffleBlockResolver.writeIndexFileAndCommit(shuffleId, mapId, partitionLengths, tmp);
+      } finally {
+        if (tmp.exists() && !tmp.delete()) {
+          logger.error("Error while deleting temp file {}", tmp.getAbsolutePath());
+        }
       }
     }
     mapStatus = MapStatus$.MODULE$.apply(blockManager.shuffleServerId(), partitionLengths);
   }
 
+  private long[] writeToExternalShuffleWriter(
+          SerializerInstance serInstance,
+          Iterator<Product2<K, V>> records) throws IOException {
+    long[] pLengths = new long[numPartitions];
+    ShufflePartitionObjectWriter writer = new ShufflePartitionObjectWriter(
+      fileBufferSize, serInstance, appId, mapId, shuffleId, pluggableWriteSupport);
+    while (records.hasNext()) {
+      final Product2<K, V> record = records.next();
+      final K key = record._1();
+      final int partition = partitioner.getPartition(key);
+      // TODO: Buffer those of same key, maybe?
+      writer.startNewPartition(partition);
+      try {
+        writer.write(key, record._2());
+        pLengths[partition] = writer.commitCurrentPartition();
```

Comment:
...Primarily because we don't want to commit prematurely. I wonder if we should keep a running set of open writers per partition and then commit all the things at the end.